### ISSUE TITLE
Generic function with default second type parameter -> Debug Failure. False expression.

### ISSUE THREAD

**Issue by mpal9000** [Microsoft/TypeScript#18938]
**TypeScript Version:**  2.5.3

**Code**

```ts
interface ObjA {
  y?:string,
  // [key:string]:any, // correctly fails to compile with error "... and 'y' does not exist in type 'MyObjA'."
}

interface ObjB {[key:string]:any}

interface Opts<A, B> {a:A, b:B}

const fn = <
  A extends ObjA,
  // A, // correctly fails to compile with error "... and 'y' does not exist in type 'MyObjA'."
  B extends ObjB = ObjB
  // B extends ObjB // correctly fails to compile with error "Type 'MyObjA' has no properties in common with type 'ObjA'.", when `fn` is called with `fn<MyObjA, ObjB>(...`
>(opts:Opts<A, B>):string => 'Z'

interface MyObjA {
  x:string,
  // y:string, // correctly compiles without errors
}

fn<MyObjA>({
// fn({ // correctly compiles without errors
// fn<MyObjA, ObjB>({ // correctly fails to compile with error "Type 'MyObjA' has no properties in common with type 'ObjA'."
  a: {x: 'X', y: 'Y'},
  b: {},
})
```

**Expected behavior:**
Fail to compile with error "Type 'MyObjA' has no properties in common with type 'ObjA'."

**Actual behavior:**
"Error: Debug Failure. False expression."

**Comment by mpal9000** [comment]
One more case with type mismatch on property `z`

```ts
interface ObjA {
  y?:string,
  z?:never,
  [key:string]:any,
}

interface ObjB {[key:string]:any}

interface Opts<A, B> {a:A, b:B}

const fn = <
  A extends ObjA,
  // A, // correctly compiles without errors
  B extends ObjB = ObjB
  // B extends ObjB // correctly fails to compile with error "Type 'number' is not assignable to type 'never'.", when `fn` is called with `fn<MyObjA, ObjB>(...`
>(opts:Opts<A, B>):string => 'Z'

interface MyObjA {
  x:string,
  y:string,
  z?:number,
}

fn<MyObjA>({
// fn({ // correctly compiles without errors
// fn<MyObjA, ObjB>({ // correctly fails to compile with error "Type 'number' is not assignable to type 'never'."
  a: {x: 'X', y: 'Y'},
  b: {},
})
```

**Comment by sandersn** [comment]
This happens because the error reporting path forgets to call `fillMissingTypeArguments`, unlike the success path.

**Comment by sandersn** [comment]
Fix is up at #19112
